Cannot use `or {}` block nor `?` with interface that contains Optional fields.

### Describe the bug

`or {}` block fails if used against an interface with cryptic error.

### Expected Behavior

Expected to be able to handle Option interface value.

### Current Behavior

```
==================
/tmp/v_1000/main.12350850225293409791.tmp.c:12184: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
module main

interface Greeting {
	tt ?string
}

struct Hello {
	tt    ?string
	value string
}

struct Hi {
	tt ?string = none
}

fn greet(g Greeting) {
       // Problematic piece of code here
       // If I leave it unchecked, the compiler complains (as expected)
	t := g.tt or { "UNKNOWN" }
	println(t)
}

greet(Hello{})
greet(Hi{})
```

### Possible Solution

_No response_

### Additional Information/Context

Might be a duplicate of #14324.

### V version

0.3.2

### Environment details (OS name and version, etc.)
```
OS: Linux version 6.0.9-300.fc37.x86_64 
CC: cc (GCC) 12.2.1 20220819 (Red Hat 12.2.1-2) 
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
```

It works if I use a match-case to identify type & dereference underlying type.

```v
module main

interface Greeting {
	tt ?string
}

struct Hello {
	tt ?string = none
	value string
}

struct Hi {
	tt ?string = none
}

fn greet(g Greeting) {
	match g {
		Hi {
			t := (*g).tt or {
				"hi"
			}
			println(t)
		}
		Hello {
			t := (*g).tt or {
				"hello"
			}
			println(t)
		}
		else {
			panic("unknown")
		}
	}
}

greet(Hi{})
greet(Hi{tt: "new hi"})
greet(Hello{})
greet(Hello{tt: "yeah"})
```

If i don't dereference `g` I get the following error 
```
/tmp/v_1000/main.15271677000846057392.tmp.c:12180: error: struct or union expected
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

Also, structs implementing the interface must have a default `none` value on the Optional field, or else it prints `println(NIL)` when `tt` isn't set.
With `Hi` being set like so :
```v
struct Hi {
	tt ?string
}
```
The output will be
```
println(NIL)
new hi
hello
yeah
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3583,10 +3583,14 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')
-		if is_opt_or_res { g.write('*(') }
+		if is_opt_or_res {
+			g.write('*(')
+		}
 		g.expr(node.expr)
 		g.write('.${node.field_name}')
-		if is_opt_or_res { g.write(')') }
+		if is_opt_or_res {
+			g.write(')')
+		}
 		g.or_block(tmp_var, node.or_block, node.typ)
 		g.write(stmt_str)
 		g.write(' ')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3583,7 +3583,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')
-		if is_opt_or_res { g.write('*)') }
+		if is_opt_or_res { g.write('*(') }
 		g.expr(node.expr)
 		g.write('.${node.field_name}')
 		if is_opt_or_res { g.write(')') }

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3577,18 +3577,18 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	}
 
 	if node.or_block.kind != .absent && !g.is_assign_lhs && g.table.sym(node.typ).kind != .chan {
-		mut is_opt_or_res := node.typ.has_flag(.optional) || node.typ.has_flag(.result)
+		is_ptr := g.table.sym(g.unwrap_generic(node.expr_type)).kind in [.interface_, .sum_type]
 		stmt_str := g.go_before_stmt(0).trim_space()
 		styp := g.typ(node.typ)
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')
-		if is_opt_or_res {
+		if is_ptr {
 			g.write('*(')
 		}
 		g.expr(node.expr)
 		g.write('.${node.field_name}')
-		if is_opt_or_res {
+		if is_ptr {
 			g.write(')')
 		}
 		g.or_block(tmp_var, node.or_block, node.typ)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3582,8 +3582,10 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')
+		g.write('*(')
 		g.expr(node.expr)
 		g.write('.${node.field_name}')
+		g.write(')')
 		g.or_block(tmp_var, node.or_block, node.typ)
 		g.write(stmt_str)
 		g.write(' ')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3577,15 +3577,16 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	}
 
 	if node.or_block.kind != .absent && !g.is_assign_lhs && g.table.sym(node.typ).kind != .chan {
+		mut is_opt_or_res := node.typ.has_flag(.optional) || node.typ.has_flag(.result)
 		stmt_str := g.go_before_stmt(0).trim_space()
 		styp := g.typ(node.typ)
 		g.empty_line = true
 		tmp_var := g.new_tmp_var()
 		g.write('${styp} ${tmp_var} = ')
-		g.write('*(')
+		if is_opt_or_res { g.write('*)') }
 		g.expr(node.expr)
 		g.write('.${node.field_name}')
-		g.write(')')
+		if is_opt_or_res { g.write(')') }
 		g.or_block(tmp_var, node.or_block, node.typ)
 		g.write(stmt_str)
 		g.write(' ')

--- a/vlib/v/tests/struct_selector_or_block_test.v
+++ b/vlib/v/tests/struct_selector_or_block_test.v
@@ -1,0 +1,26 @@
+module main
+
+interface Greeting {
+	tt ?string
+}
+
+struct Hello {
+	tt    ?string = none
+	value string
+}
+
+struct Hi {
+	tt ?string = none
+}
+
+fn greet(g Greeting) string {
+	// Problematic piece of code here
+	// If I leave it unchecked, the compiler complains (as expected)
+	t := g.tt or { 'UNKNOWN' }
+	return t
+}
+
+fn test_main() {
+	assert greet(Hello{}) == 'UNKNOWN'
+	assert greet(Hello{'cool', ''}) == 'cool'
+}

